pkg/shell/nav.tsx

```diff
@@ -60,60 +117,59 @@ export class CockpitNav extends React.Component {
         };
 
         this.clearSearch = this.clearSearch.bind(this);
+        this.props = props;
     }
 
     componentDidMount() {
-        const self = this;
         const sel = this.props.selector;
         // Click on active menu item (when using arrows to navigate through menu)
         function clickActiveItem() {
             const cur = document.activeElement;
-            if (cur.nodeName === "INPUT") {
-                const el = document.querySelector("#" + sel + " li:first-of-type a");
+            if (cur instanceof HTMLInputElement) {
+                const el: HTMLElement | null = document.querySelector("#" + sel + " li:first-of-type a");
```

Comment:
I didn't find any extensive documentation of how you are supposed to use this.  

There are overloads that allow TS to infer precise result types for certain trivial selectors: `querySelector("a")` has type `HTMLAnchorElement`, for example.

The general form is `querySelector<E extends Element = Element>(sel: string): E | null` and it can either absorb the type from the environment (like here, where TS infers `E = HTMLElement` from the type of `el`), _or_ we can specify `E` explicitly like below and let TS infer the type of `e`.

I have no idea which is the intended use. But I agree that it feels more natural to let the types flow from expressions outwards, and not from declarations inwards. So let's use `querySelector<HTMLElement>`.

Reply:
I don't have a strong opinion which way around. Both feel like a "developer promise" to me, as without careful inspection of the selector (which is dynamic even) tsc only knows `Element`. But I sympathize with `querySelector<HTMLElement>` as well.